ray-on-gke/cloudbuild.yaml

```diff
@@ -63,31 +69,88 @@ steps:
 
         ray job submit --working-dir ../example_ray_job_scripts \
         --address=http://127.0.0.1:8265 -- python ray_job.py \
-        || ( echo "false" >> /workspace/ray_result.txt )
+        || ( exit 1 )
+        echo "pass" > /workspace/ray_result.txt
+    allowExitCodes: [1]
+    waitFor: ['create gke cluster']
 
-        terraform destroy -var=project_id=$PROJECT_ID \
+  - id: 'test jupyterhub'
+    name: 'gcr.io/$PROJECT_ID/terraform'
+    entrypoint: 'bash'
+    args:
+      - '-c'
+      - |
+        cd jupyter-on-gke/tests
+        python3 change_jupyter_config.py
+
+        cd ../
+        terraform apply \
+        -var=project_id=$PROJECT_ID \
         -var=namespace=$SHORT_SHA \
-        -var=service_account=$_USER_NAME-$SHORT_SHA-system-account \
-        -auto-approve -no-color
-    dir: 'ray-on-gke/user/'
-    waitFor: ['get kube config']
+        -var=add_auth=false \
+        -var=enable_iap_service=false \
+        -auto-approve -no-color \
+        || ( exit 1 )
+        echo "pass" > /workspace/jupyterhub_tf_result.txt
+
+        kubectl wait --all pods -n $SHORT_SHA --for=condition=Ready --timeout=300s
+        kubectl get services -n $SHORT_SHA
+        kubectl get service proxy-public -n $SHORT_SHA --output jsonpath='{.status.loadBalancer.ingress[0].ip}' > /workspace/jupyterhub_host_url.txt
+        echo "HOST URL is " $(cat /workspace/jupyterhub_host_url.txt)
+
+        cd tests/
+        python3 test_hub.py $(cat /workspace/jupyterhub_host_url.txt) || ( exit 1 )
+        echo "pass" > /workspace/jupyterhub_test_result.txt 
+    allowExitCodes: [1]
+    waitFor: ['test ray cluster']
 
   - id: 'clean gke cluster'
-    name: 'gcr.io/cloud-builders/gcloud'
+    name: 'gcr.io/$PROJECT_ID/terraform'
     entrypoint: 'bash'
     args:
       - '-c'
       - |
-        gcloud container clusters delete ray-$SHORT_SHA-$_PR_NUMBER-cluster --zone $_ZONE  --project $PROJECT_ID --quiet
+        cd jupyter-on-gke/
+        terraform destroy \
+        -var=project_id=$PROJECT_ID \
+        -var=namespace=$SHORT_SHA \
+        -var=add_auth=false \
+        -var=enable_iap_service=false \
+        -auto-approve -no-color 
+
+        cd ../ray-on-gke/user/
+        terraform destroy -var=project_id=$PROJECT_ID \
+        -var=namespace=$SHORT_SHA \
+        -var=service_account=$_USER_NAME-$SHORT_SHA-system-account \
+        -auto-approve -no-color
+
+        gcloud container clusters delete ray-$SHORT_SHA-$_PR_NUMBER-cluster --zone $_ZONE  --project $PROJECT_ID --quiet 
+
+        if [[ $(cat /workspace/gke_cluster_result.txt) != "pass" ]]; then
```

Comment:
Is it possible to put these checks into a separate build stage instead of the clean up one? This way you can set `allowFailure: true` in the other stages (instead of allowExitCodes) except the final stage that checks these values

Reply:
(and it's a little misleading to see the failure in the clean up stage since it won't be immediately clear the build failed due to the test and not due to the clean up)

Reply:
done